### ISSUE TITLE
Format "wvtt" of WEBVTT subtitle muxed of mp4box won't show subtitle.

### ISSUE THREAD

**Issue by MartinEesmaa** [gpac/gpac#2243]
- [X] I looked for a similar issue and couldn't find any.
- [X] I tried with the latest version of GPAC. Installers available at http://gpac.io/downloads/gpac-nightly-builds/
- [X] I give enough information for contributors to reproduce my issue (meaningful title, github labels, platform and compiler, command-line ...). I can share files anonymously with this dropbox: https://www.mediafire.com/filedrop/filedrop_hosted.php?drop=eec9e058a9486fe4e99c33021481d9e1826ca9dbc242a6cfaab0fe95da5e5d95

Hello. I have a problem about muxing WEBVTT subtitle to mp4box. On muxed mp4 video using mp4box with WEBVTT won't show the subtitle only in mpv or VLC.
In mkvtoolnix muxed mkv with webvtt, it shows correct subtitle.
See the two screenshots and shared files below:

WEBVTT texts:
```
WEBVTT

00:00.000 --> 00:02.000
Test
```

OS: Windows 11 Pro 21H2

mp4box version: 2.1-DEV-rev259-gfc4749f9-master

mp4box + mkvtoolnix muxed file: https://mega.nz/folder/nxVlgKYZ#HMLcZzKwtVDgIpSAAG6h3Q

mkvtoolnix muxed:
![mkvtoolnixmuxed](https://user-images.githubusercontent.com/88035011/185731203-1c4c1749-4d7e-42dc-a6d4-a533c998349f.png)

mp4box muxed:
![mp4boxmuxed](https://user-images.githubusercontent.com/88035011/185731221-67a3fef3-2172-4c7b-b0eb-6461f63bb817.png)

- Martin Eesmaa

**Comment by vt-idiot** [comment]
# TL;DR You _should_ use tx3g for subtitles when output to an .mp4 file for local playback!

Luckily, GPAC has a [filter](https://github.com/gpac/gpac/wiki/vtt2tx3g) for this that you can use like so:
* `-add subtitles.vtt:@vtt2tx3g`
    * note the `:@vtt2tx3g` at the end of the filename

You can optionally define `fontname` and `fontsize` if you want, like this:
* `-add subtitles.vtt:@vtt2tx3g:fontname="Bebas Neue":fontsize=12`
    * `fontname` is a string and thus must be in quotes (on Windows at least)
    * `fontsize` is an unsigned integer, defaults to 18 if unspecified, and no quotes are needed

<details>
  <summary>More on tx3g style options, as many players seem to ignore them</summary>

<br>

Do note, that seemingly _everything_ ignores those two options, but it technically doesn't hurt if you specify them.
* GPAC player will honor both settings
* I don't have QuickTime installed anymore (neither should you on Windows) or a Mac to check on, but I would not be surprised if it honored them
* MPC-HC will always use its own "default style" for tx3g
* MPV will ignore them, even if the config file does not have subtitle font settings
* VLC also ignores them and uses its Subtitle/OSD font settings for any text subtitles that are not ASS/SSA, and for SRT when it does not specify a style
* ffplay was the only thing other than GPAC's player that seemed to honor them
    * That might be because I compiled it with `--enable-libass --enable-fontconfig --enable-libfribidi`:
```
ffplay -vf "subtitles=Test.MP4Box.Bebas.mp4" -i Test.MP4Box.Bebas.mp4 -hide_banner
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'Test.MP4Box.Bebas.mp4':0
  Metadata:
    major_brand     : isom
    minor_version   : 1
    compatible_brands: isomavc1
    creation_time   : 2022-08-22T18:27:26.000000Z
    encoder         : GPAC-2.1-DEV-rev173-g82dd6d2b-master
  Duration: 00:07:06.43, start: 0.000000, bitrate: 4045 kb/s
  Stream #0:0[0x1](und): Video: h264 (High) (avc1 / 0x31637661), yuv420p(tv, bt709, progressive), 1480x1080, 3860 kb/s, 23.98 fps, 23.98 tbr, 96k tbn (default)
    Metadata:
      creation_time   : 2022-08-22T18:27:26.000000Z
      handler_name    : video.h264@GPAC2.1-DEV-rev173-g82dd6d2b-master
      vendor_id       : [0][0][0][0]
  Stream #0:1[0x2](und): Audio: ac3 (ac-3 / 0x332D6361), 48000 Hz, mono, fltp, 192 kb/s (default)
    Metadata:
      creation_time   : 2022-08-22T18:27:30.000000Z
      handler_name    : ac3@GPAC2.1-DEV-rev173-g82dd6d2b-master
      vendor_id       : [0][0][0][0]
    Side data:
      audio service type: main
  Stream #0:2[0x3](und): Subtitle: mov_text (tx3g / 0x67337874), 1480x1080, 0 kb/s (default)
    Metadata:
      creation_time   : 2022-08-22T18:27:30.000000Z
      handler_name    : vtt:@vtt2tx3g:fontname=Bebas Neue:fontsize=12@GPAC2.1-DEV-rev173-g82dd6d2b-master
[Parsed_subtitles_0 @ 000001d9eda64980] libass API version: 0x1600000
[Parsed_subtitles_0 @ 000001d9eda64980] libass source: commit: 0.16.0-1-ga2b39cde4ecb74d5e6fccab4a5f7d8ad52b2b1a4
[Parsed_subtitles_0 @ 000001d9eda64980] Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 4.3.0 (COMPLEX)
[Parsed_subtitles_0 @ 000001d9eda64980] Using font provider directwrite (with GDI)
[Parsed_subtitles_0 @ 000001d9eda64680] libass API version: 0x1600000
[Parsed_subtitles_0 @ 000001d9eda64680] libass source: commit: 0.16.0-1-ga2b39cde4ecb74d5e6fccab4a5f7d8ad52b2b1a4
[Parsed_subtitles_0 @ 000001d9eda64680] Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 4.3.0 (COMPLEX)
[Parsed_subtitles_0 @ 000001d9eda64680] Using font provider directwrite (with GDI)
[Parsed_subtitles_0 @ 000001d9eda63680] libass API version: 0x1600000
[Parsed_subtitles_0 @ 000001d9eda63680] libass source: commit: 0.16.0-1-ga2b39cde4ecb74d5e6fccab4a5f7d8ad52b2b1a4
[Parsed_subtitles_0 @ 000001d9eda63680] Shaper: FriBidi 1.0.12 (SIMPLE) HarfBuzz-ng 4.3.0 (COMPLEX)
[Parsed_subtitles_0 @ 000001d9eda63680] Using font provider directwrite (with GDI)
[Parsed_subtitles_0 @ 000001d9eda63680] fontselect: (Bebas Neue, 400, 0) -> BebasNeue-Regular, 0, BebasNeue-Regular
  52.96 A-V: -0.041 fd= 243 aq=  642KB vq=14718KB sq=    0B f=0/0
```
![image](https://user-images.githubusercontent.com/81622808/185994228-794bb69d-becf-4941-ba84-ea8822b38f67.png)
</details>

Anyways, mp4box is doing exactly what you're asking it to, the problem is that the resulting output is not something any player I am aware of will know how to deal with. VLC can actually see the wvtt subtitle track, but it does not play it back. MPV and MPC-HC (LAV Filters) can't see it at all, likely because `libavformat` doesn't recognize it as a text stream.

---

`mp4box -add video.h264 -add audio.ac3 -add subtitles.vtt -inter 500 -tight --forcesync Test.MP4Box.wvtt.mp4`
* Subtitles don't seem to work anywhere other than GPAC player. ffplay can't see the subtitles when you try to ask it to either, more on that later.
```

Track Importing MPEG-4 AVC - Width 1480 Height 1080 FPS 96000/4004
AVC|H264 Import results: 10201 samples (10373 NALUs) - Slices: 59 I 3027 P 7116 B - 57 SEI - 1 IDR - 0 CRA
AVC|H264 Stream uses forward prediction - stream CTS offset: 10 frames
Track Importing AC-3 - SampleRate 48000 Num Channels 1
Track Importing WebVTT - Text track 1480 x 1080 layer 0
Full Interleaving
```

`mp4box -add video.h264 -add audio.ac3 -add subtitles.vtt:@vtt2tx3g -inter 500 -tight --forcesync Test.MP4Box.mp4`
* Subtitles will work
```
Track Importing MPEG-4 AVC - Width 1480 Height 1080 FPS 96000/4004
AVC|H264 Import results: 10201 samples (10373 NALUs) - Slices: 59 I 3027 P 7116 B - 57 SEI - 1 IDR - 0 CRA
AVC|H264 Stream uses forward prediction - stream CTS offset: 10 frames
Track Importing AC-3 - SampleRate 48000 Num Channels 1
Track Importing Timed Text - Text track 1480 x 1080 font Serif (size 18) layer 0
Full Interleaving
```

`mp4box -add video.h264 -add audio.ac3 -add subtitles.vtt:@vtt2tx3g:fontname="Bebas Neue":fontsize=12 -flat Test.MP4Box.Bebas.mp4`
* Same result as above, with ffplay being the only thing other than GPAC player that seemingly honors the font and size.
```
Track Importing MPEG-4 AVC - Width 1480 Height 1080 FPS 96000/4004
OpenGOP detected - adjusting file brand
AVC|H264 Import results: 10201 samples (10373 NALUs) - Slices: 59 I 3027 P 7116 B - 57 SEI - 1 IDR - 0 CRA
AVC|H264 Stream uses forward prediction - stream CTS offset: 10 frames
Track Importing AC-3 - SampleRate 48000 Num Channels 1
Track Importing Timed Text - Text track 1480 x 1080 font Bebas Neue (size 12) layer 0
Flat storage
```

This isn't unique to GPAC either, ffmpeg simply does nothing with vtt subtitles when muxing to mp4 if you don't force it to convert to tx3g.

<details>
<summary>Muxing subtitles into MP4 with ffmpeg</summary>

<br>

`ffmpeg -i video.h264 -i audio.ac3 -i subtitles.vtt -c copy -hide_banner Test.mp4`

* ffmpeg will do nothing with the subtitles, so the output will have none

```
Input #0, h264, from 'video.h264':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Video: h264 (High), yuv420p(tv, bt709, progressive), 1480x1080, 23.98 fps, 23.98 tbr, 1200k tbn
[ac3 @ 00000176fa7a97c0] Estimating duration from bitrate, this may be inaccurate
Input #1, ac3, from 'audio.ac3':
  Duration: 00:07:06.43, start: 0.000000, bitrate: 192 kb/s
  Stream #1:0: Audio: ac3, 48000 Hz, mono, fltp, 192 kb/s
Input #2, webvtt, from 'subtitles.vtt':
  Duration: N/A, bitrate: N/A
  Stream #2:0: Subtitle: webvtt
[mp4 @ 00000176fa7a0a80] track 1: codec frame size is not set
Output #0, mp4, to 'Test.mp4':
  Metadata:
    encoder         : Lavf59.24.100
  Stream #0:0: Video: h264 (High) (avc1 / 0x31637661), yuv420p(tv, bt709, progressive), 1480x1080, q=2-31, 23.98 fps, 23.98 tbr, 1200k tbn
  Stream #0:1: Audio: ac3 (ac-3 / 0x332D6361), 48000 Hz, mono, fltp, 192 kb/s
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
  Stream #1:0 -> #0:1 (copy)
```
---
`ffmpeg -i video.h264 -i audio.ac3 -i subtitles.vtt -c:v copy -c:a copy -c:s mov_text -hide_banner Test.tx3g.mp4`
* note **-c:s mov_text**
* you can now see the following additional line in the output:
    * `Stream #2:0 -> #0:2 (webvtt (native) -> mov_text (native))`
* the file has working subtitles
</details>

---

Is MP4Box doing it wrong? No idea, **but I don't think so**. [It _will_ put .vtt directly into ISOBMFF if you ask it to](https://github.com/gpac/gpac/wiki/Subtitling-with-GPAC), since there are scenarios where you might need it - but they're all DASH or HLS specific as far as I am aware. You would have to double check the contents of **ISO/IEC 14496-30:2018** to see if GPAC is handling it "wrong", but it's not available to read freely in its entirety.


<details>
<summary>ffprobe vs mp4box -info</summary>

ffprobe seems to think the embedded wvtt subtitles are a **Data** stream. That is probably because it isn't recognizing them, `mp4box -info` _correctly_ identifies the stream as **Text**. If this is an issue with ffmpeg (specifically, libavformat - and I suspect it **is** an issue with libavformat) and how it handles WebVTT within ISOBMFF files, then that would explain why nothing can play it - practically everything uses libavformat or other ffmpeg libaries in some form or another, including mpv.

`ffprobe -i Test.MP4Box.wvtt.mp4 -hide_banner`
```
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'Test.MP4Box.wvtt.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 1
    compatible_brands: isomavc1
    creation_time   : 2022-08-22T17:44:59.000000Z
    encoder         : GPAC-2.1-DEV-rev173-g82dd6d2b-master
  Duration: 00:07:06.43, start: 0.000000, bitrate: 4048 kb/s
  Stream #0:0[0x1](und): Video: h264 (High) (avc1 / 0x31637661), yuv420p(tv, bt709, progressive), 1480x1080, 3860 kb/s, 23.98 fps, 23.98 tbr, 96k tbn (default)
    Metadata:
      creation_time   : 2022-08-22T17:44:59.000000Z
      handler_name    : video.h264@GPAC2.1-DEV-rev173-g82dd6d2b-master
      vendor_id       : [0][0][0][0]
  Stream #0:1[0x2](und): Audio: ac3 (ac-3 / 0x332D6361), 48000 Hz, mono, fltp, 192 kb/s (default)
    Metadata:
      creation_time   : 2022-08-22T17:45:02.000000Z
      handler_name    : ac3@GPAC2.1-DEV-rev173-g82dd6d2b-master
      vendor_id       : [0][0][0][0]
    Side data:
      audio service type: main
  Stream #0:2[0x3](und): Data: none (wvtt / 0x74747677), 0 kb/s (default)
    Metadata:
      creation_time   : 2022-08-22T17:45:02.000000Z
      handler_name    : vtt@GPAC2.1-DEV-rev173-g82dd6d2b-master
Unsupported codec with id 0 for input stream 2
```

`mp4box -info Test.MP4Box.wvtt.mp4`
```
# Movie Info - 3 tracks - TimeScale 600
Duration 00:07:06.431
Fragmented: no
Progressive (moov before mdat)
Major Brand isom - version 1 - compatible brands: isom avc1
Created: GMT Mon Aug 22 17:44:59 2022

File has root IOD (9 bytes)
Scene PL 0xff - Graphics PL 0xff - OD PL 0xff
Visual PL: ISO Reserved Profile (0x15)
Audio PL: No audio capability required (0xff)
No streams included in root OD

iTunes Info:
        tool: GPAC-2.1-DEV-rev173-g82dd6d2b-master

# Track 1 Info - ID 1 - TimeScale 96000
Media Duration 00:07:05.633
Track has 1 edits: track duration is 00:07:05.466
Track flags: Enabled In Movie In Preview
Media Info: Language "Undetermined (und)" - Type "vide:avc1" - 10201 samples
Visual Sample Entry Info: width=1480 height=1080 (depth=24 bits)
Visual Track layout: x=0 y=0 width=1480 height=1080
AVC/H264 Video - Visual Size 1480 x 1080
        AVC Info: 1 SPS - 1 PPS - Profile High @ Level 4.1
        NAL Unit length bits: 32
        Chroma format YUV 4:2:0 - Luma bit depth 8 - chroma bit depth 8
        SPS#1 hash: 4600C7EB8CB28446D56584DD75C796F480CC87F4
        PPS#1 hash: 09F88915DD329F83E58F6765D2073022BC4D3518
        RFC6381 Codec Parameters: avc1.640029
        Average GOP length: 182 samples
        Max sample duration: 4004 / 96000

# Track 2 Info - ID 2 - TimeScale 48000
Media Duration 00:07:06.432
Track flags: Enabled In Movie In Preview
Media Info: Language "Undetermined (und)" - Type "soun:ac-3" - 13326 samples
Alternate Group ID 1
        AC-3 stream - Sample Rate 48000 - 1 channel(s) - bitrate 192000
        RFC6381 Codec Parameters: ac-3
        All samples are sync
        Max sample duration: 1536 / 48000

# Track 3 Info - ID 3 - TimeScale 1000
Media Duration 00:07:02.826
Track flags: Enabled In Movie In Preview
Media Info: Language "Undetermined (und)" - Type "text:wvtt" - 130 samples
Unknown Text Stream
        Size 1480 x 1080 - Translation X=0 Y=0 - Layer 0
        RFC6381 Codec Parameters: wvtt
        All samples are sync
        Max sample duration: 35827 / 1000
```

---


Of course, GPAC itself can play the files with wvtt subtitles within them just fine.
```
gpac -gui Test.MP4Box.wvtt.mp4
Refreshing all options registry, this may take some time ... done
No accelerated IMDCT transform found
[JS] WebVTT Rendered Loaded
```
Result:
![image](https://user-images.githubusercontent.com/81622808/185998338-d8b4e8c8-e3f1-4dca-ba47-a2d9e3418285.png)

It also honors font information for tx3g subtitles, although ttxtdec seems to crap its pants if you seek in the video at all:
```
gpac -gui Test.MP4Box.Bebas.mp4
No accelerated IMDCT transform found
```
Result:
![image](https://user-images.githubusercontent.com/81622808/185999470-4db8cfe7-64d4-4668-9b63-b5c9c1a2d3c8.png)
</details>

<details>
<summary>MKV =! MP4</summary>

<br>

Media players and demuxers handle the two very differently, because they're muxed very _very_ differently. Don't forget that the ISOBMFF (.mp4) container is originally based off QuickTime's **.MOV** format. Atoms, for example, were defined by the QuickTime container more than 30 years ago! Where do you think the `moov` header got its name? MacOS defined the type code for QuickTime files as `MooV`.  Changes to what formats it supports, and how it supports them, are basically at ISO/MPEG's whim as well.

MKV was built from the ground up to be a lot more versatile (and **free/open...**) as a container when it comes to supported formats, and you can do many _many_ things with MKV that might not work as expected with ISOBMFF, like FLAC audio, things that simply do not work at all with ISOBMFF, like ASS/SSA subtitles, or things that MPEG expects you to use a program or transport stream for instead, like .sup Blu-Ray subtitles. And those are just the examples I can think of off the top of my head.
</details>

![image](https://c.tenor.com/ERa47gPVan8AAAAC/the-more-you-know.gif)

**Comment by MartinEesmaa** [comment]
Okay @vt-idiot, thank you! I will close this issue.
